### PR TITLE
Feature/granularity feature selection tests mm

### PR DESCRIPTION
**Pull request by malgosiam2**

### CODE REVIEW

**Review by szostkawron — COMMENTED** [review]
Great work overall! However, there are several things that should be addressed:
1. Duplicate test names: There are two tests named test_fit_and_transform. Please rename one of them to avoid confusion and potential issues with test runners.
2. Input validation tests: In the test that checks for errors with invalid inputs, consider adding more examples with different data types (e.g., lists, scalars, mixed types) to ensure all invalid cases are properly handled.
3. Inconsistent columns between fit and transform: It would be helpful to add a test that checks what happens if the X passed to transform() has different columns than the one used in fit().
4. Missing values in X: Please add a test case to verify the behavior when X contains missing values (NaN). This is a common real-world scenario and should be explicitly handled or documented.
5. Flexible input types for X: Currently, your implementation accepts only pd.DataFrame as input. You could consider using the existing check_input_dataset() utility from utils, which allows multiple input types and handles conversion to DataFrame. This would simplify input validation and reduce code duplication.
6. Unsupervised mode (y=None): Since fit() now supports y=None, please add a test to confirm that this mode works as expected and doesn't break in transform().

**Review by szostkawron — COMMENTED** [review]
Great, but 
1. Rename the class: change the name to the shorter version in Notion.
2. fit/transform design issue: currently, fit() and transform() must be called on the exact same data. This is not correct for a scikit-learn-style transformer.
    - fit() should learn which features to keep (store them in self.S_opt).
    - transform() should then only drop the unnecessary columns — it should not recompute the model or run accuracy checks again.
3. Spelling: There is a spelling mistake in _granual_consistency_of_B_subset method, it should be "granular"
4. Missing tests for private methods: create simple tests just to verify that they execute and return valid outputs
5. To check if the class was fitted before transform use the code:
    from sklearn.utils.validation import check_is_fitted
    check_is_fitted(self, attributes=["X_train_", "imputer_", "centers_", "u_", "optimal_c_", "np_rng_"])
6. Add test for d > number_of_features
7. Input validation tests: extend input validation coverage with more data type variants — text, DataFrame, NumPy array, Python list, etc. — to ensure all invalid or edge-case inputs are handled consistently.

**Comment by malgosiam2** [comment]
All changes were made

**Review by szostkawron — CHANGES_REQUESTED** [review]
Looks great, but:
1. Remove comments or translate them into clear English for consistency and readability.
2. To check if the class has been fitted before transform(), use the standard scikit-learn utility: 
check_is_fitted(self, attributes=["delta_cache_", ..., "entropy_cache_"])
3. Standardize attribute naming: for private/hidden attributes, use the self.name_ convention consistently. Currently, underscores appear both before and after attribute names.
4. Add a test case where d > number_of_features to ensure the selector behaves correctly.
5. Reconsider whether transform() should convert categorical features into numerical values. Typically, it should return the same dataset with only the selected columns, leaving data types unchanged.
6. Some checks in transform() are redundant. For example, verifying that the selector is fitted both before and after checking self.S is unnecessary.
7. Review the usage of D[0]: currently, it is sometimes used as a column index and sometimes as a column name. This inconsistency could lead to bugs.
8. Verify that transform() works correctly on a DataFrame with a single row.
9. Expand the sample datasets in the tests with additional scenarios to ensure coverage (e.g., more mixed numerical/categorical features, edge cases).
10. Add a test to ensure that the number of columns in transform(X) is always less than or equal to d.
11. Check the behavior of transform() when X has only a single column, both numerical and categorical, to ensure it handles edge cases.
12. Consider optimizing performance: some loops can be vectorized for speed, especially in methods like _calculate_delta_for_column_subset.

**Comment by malgosiam2** [comment]
1) done
2) done
3) done 
4) done
5) method adjustemnts were made - now transform does not change categorical to numeric
6) done 
7) Now we always use column name instead of col index 
8) done
9) there is already enough tests for that
10) done
11) done for numerical
12) it is aready quite optimised - It was much worse

**Review by szostkawron — COMMENTED** [review]
Almost done — just one issue regarding attribute naming.
Attributes created inside fit() should follow the scikit-learn convention and end with an underscore, e.g. delta_cache_ and entropy_cache_, rather than _delta_cache or _entropy_cache.
Additionally, in the test_transform_without_fit test, the check_is_fitted call should include all attributes that are created in fit(), not just a subset. This ensures that the fitted-state check is complete and consistent.

**Comment by malgosiam2** [comment]
ok

**Review by szostkawron — APPROVED** [review]
